test/D2L.Services.Core.Extensions.UnitTests/ParallelizationExtensionsTests.cs

```diff
@@ -0,0 +1,153 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using D2L.Services;
+using NUnit.Framework;
+
+namespace D2L.Services {
+	
+	[TestFixture]
+	[Category( "Unit" )]
+	internal sealed class ParallelizationExtensionsTests {
+		
+		private Random m_rng;
+		
+		[TestFixtureSetUp]
+		public void TestFixtureSetUp() {
+			m_rng = new Random( DateTime.Now.GetHashCode() );
+		}
+		
+		[Test]
+		public async Task FilterInParallelAsyncTest() {
+			int[] numbers = new[]{ 1, 6, 1, 8, 0, 3, 3, 9, 8, 8, 7, 5 };
+			int[] oddOnly = new[]{ 1,    1,       3, 3, 9,       7, 5 };
+			
+			IEnumerable<int> result = await numbers.FilterInParallelAsync( async n => {
+				await RandomDelay( 900 ).SafeAsync();
+				return n % 2 != 0;
+			}).SafeAsync();
+			
+			CollectionAssert.AreEqual( oddOnly, result );
+		}
+		
+		[Test]
+		public async Task MapInParallelAsyncTest() {
+			int[] numbers = new[]{ 3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5, 8, 9, 7, 9 };
+			
+			IEnumerable<int> expected = numbers.Select( n => (n + 5) % 10 );
+			IEnumerable<int> actual = await numbers.MapInParallelAsync( async n => {
+				await RandomDelay( 900 ).SafeAsync();
+				return ( n + 5 ) % 10;
+			}).SafeAsync();
+			
+			CollectionAssert.AreEqual( expected, actual );
+		}
+		
+		[Test]
+		public async Task ForEachInParallelAsync_ExceptionHandlingTest() {
+			int[] numbers = new[]{ 0, 0, 3, 0, 1, 2 };
+			Exception[] expectedInnerExceptions = new TestException[]{
+				new TestException( 3 ),
+				new TestException( 1 ),
+				new TestException( 2 )
+			};
+			
+			Task forEachTask = numbers.ForEachInParallelAsync( async n => {
+				await RandomDelay( 900 ).SafeAsync();
+				if( n != 0 ) {
+					throw new TestException( n );
+				}
+			});
+			
+			try {
+				await forEachTask.SafeAsync();
+			} catch( AggregateException exception ) {
+				CollectionAssert.AreEqual( expectedInnerExceptions, exception.InnerExceptions );
+				Assert.Pass();

```

Comment:
I did it this way so that I can be sure that the exception is thrown at the correct location. If the exception were thrown when the Task was created instead of on the `await`-- which would indicate an error, then the pass would pass if I simply used [ExpectedException], and all assertions past that point would be skipped.